### ISSUE TITLE
[sampling] top_k>128 returns token 0 for every seed when all logits are negative: signed atomicMax on a float bit pattern

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1305]
## Summary

When every logit is negative, the `top_k > 128` sampling path returns token 0 for every seed, regardless of the distribution. Adding a constant to all logits — which softmax cannot see — makes it work. The global-max reduction uses a **signed** `atomicMax` on the float's raw bit pattern, and that ordering is inverted for negative floats, so the `-FLT_MAX` sentinel is never replaced.

## Repro

Commit: `ad067d76` (v0.23.0, `main`)
Command: `bash loop/repro/BUG-2.sh` — reproduced 2/2 from fresh processes. Or directly:

```bash
docker run --rm --gpus all -v $PWD:/src -w /src/build-dev imp:toolchain \
  ./test-compute --gtest_filter='SamplingTest.TopPTruncatesCubPath*'
```

Environment: RTX 5090 (sm_120a), WSL2, CUDA 13.3.1, `build-dev` Release.

The two tests feed the **same distribution**, V=512, `top_k=200`, `top_p=0.75`, `temperature=1.0`, 200 seeds:

* `TopPTruncatesCubPath` — logits are `ln(p)`, so the maximum is `-0.92` and every logit is negative.
* `TopPTruncatesCubPathShifted` — logits are `ln(p) + 5`, so the maximum is `+4.08`.

Softmax is shift-invariant, so these must behave identically.

## Expected vs Actual

Expected: both draw from `{137, 42, 301}` (the `top_p=0.75` nucleus, computed in-test from the same probabilities) with a spread over seeds.

Actual:

```
[ RUN      ] SamplingTest.TopPTruncatesCubPath
CUB, all-negative logits: all 200 seeds drew token 0 (token 0 — the sampler's error sentinel)
[  FAILED  ] SamplingTest.TopPTruncatesCubPath
[ RUN      ] SamplingTest.TopPTruncatesCubPathShifted
[       OK ] SamplingTest.TopPTruncatesCubPathShifted (82 ms)
```

Token 0 is neither the argmax (137) nor in the nucleus. No error is logged — the failure is completely silent.

## Evidence

`src/compute/sampling_topk_topp.cu:397`, in `softmax_max_kernel`:

```cpp
atomicMax(reinterpret_cast<int*>(d_max), __float_as_int(mx));
```

with the accumulator pre-seeded to `__float_as_int(-FLT_MAX)` (`:578-580`).

For a negative float, larger magnitude ⇒ larger unsigned bit pattern ⇒ **larger** value when reinterpreted as `int32`. So among negatives, a signed `atomicMax` selects the float **minimum**, and the `-FLT_MAX` sentinel (`0xFF7FFFFF` → `-9 437 185` as `int32`) beats every ordinary negative logit (e.g. `-20.7f` → `0xC1A5D1EB` → `-1 046 326 805`). The sentinel therefore survives the reduction.

Downstream (`:600`, `softmax_to_pairs_device_kernel`) every probability becomes `expf(logit - (-FLT_MAX))` — inf, then `inf/inf` — so all sort keys are equal/NaN, the descending radix sort leaves index 0 at the head, and `topp_sample_from_sorted_kernel` (`:459`) falls through to its `chosen = sorted_indices[0]` initialiser.

The multiblock path (`top_k <= 128`) is unaffected: it reduces the max in floats via `warp_reduce_max`, no bitcast. Both multiblock variants pass.

### Second site, same defect

`src/compute/mtp_forward.cu:260`:

```cpp
__shared__ float s_block_max;
if (tid == 0) s_block_max = -1.0e30f;
__syncthreads();
atomicMax(reinterpret_cast<int*>(&s_block_max), __float_as_int(max_score));
```

These are attention scores, which are routinely negative. Same inversion, same sentinel-survives outcome, feeding `expf(s_scores[t] - gmax)` two lines later. Not separately reproduced — flagged because it is the same pattern; `rg 'atomicMax\(reinterpret_cast<int\*>'` finds exactly these two.

## Why the test suite missed it

Escape class: **E6** — the test exists and its inputs cannot trigger the fault.

`SamplerCubPath.EachCallSamplesFromItsOwnLogitsNotThePreviousCalls` (`tests/test_sampling.cu:416`) is the CUB-path test, added for #1142. It uses `h[winner] = 30.0f` with the rest at `-20.0f` — the maximum is **positive**, so the sentinel loses and the reduction is correct. It is also the only test in the suite that reaches this path.

Note the interaction with #1142: the symptom there was "the CUB path emits the same token forever". This is a second, independent mechanism with the same symptom, and it was still present after that fix.

## Severity

**S1 — silent wrong output.** Trigger condition: `top_k > SAMPLE_MAX_TOP_K` (128) **and** every logit negative at that decode step. The result is token 0 emitted repeatedly — the degeneration class of #934/#1142 — with no error, no warning, and no dependence on the sampling parameters the user set.

## Suggested test

Already written and attached to this report:

* `SamplingTest.TopPTruncatesCubPath` / `…CubPathShifted` — same distribution, logits shifted by a constant; softmax shift-invariance is the oracle. Currently **red on `main`** for the unshifted case; it is red because the product is wrong, and it should stay red until this is fixed.
* The same pair exists for the multiblock path (`TopPTruncatesMultiblockPath` / `…Shifted`), both green, which is what localises the fault to the CUB path.

A dedicated `mtp_forward` shift-invariance case would cover the second site.

Found while writing the tests for #1300; see `docs/audit/TEST_HARDENING_LOG.md`.

**Comment by kekzl** [comment]
Fixed in #1322, merged to `main`.

The `Closes #1305` keyword was present in both the PR body and the squash commit but did not fire: auto-merge performs the squash under `GITHUB_TOKEN`, and GitHub does not action closing keywords from that actor. Closing by hand — worth knowing for every PR in this repo, not just this one.
